### PR TITLE
4643 varnet

### LINKED ISSUE #4643: e2e-VarNet model
**Is your feature request related to a problem? Please describe.**
e2e-VarNet (https://arxiv.org/abs/2004.06688), a state-of-the-art MRI reconstruction model is not available in MONAI.

**Describe the solution you'd like**
Add its implementation.

**Describe alternatives you've considered**
N/A

**Additional context**

### PR DESCRIPTION
**Pull request by mersad95zd**
Fixes #4643

### Description
Implementation of the end-to-end variational network for accelerated MRI reconstruction

### Status
**ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by mersad95zd** [comment]
ToDo: Complete the unit test by fixing the issue with `torch.jit.script`

File: monai/apps/reconstruction/networks/blocks/varnetblock.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+class VarNetBlock(nn.Module):
+    """
+    A variational block based on Sriram et. al., "End-to-end variational networks for accelerated MRI reconstruction".
+    It applies data consistency and refinement to the intermediate kspace and combines those results.
+
+    Modified and adopted from: https://github.com/facebookresearch/fastMRI
+
+    Args:
+        refinement_model: the model used for refinement (typically a U-Net but can be any standard deep learning model)
```
Comment (Can-Zhao) [inline monai/apps/reconstruction/networks/blocks/varnetblock.py:28]: What is standard deep network...Maybe change it to: typically a U-Net but can be any deep learning model that XXXX
Thanks!

File: monai/apps/reconstruction/networks/blocks/varnetblock.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+            """
+            expands an image to its corresponding coil images based on the given sens_maps
+
+            Args:
+                x: image (B,1,H,W,2) with the last dimension being 2 (for real/imaginary parts)
+
+            Returns:
+                expansion of x to (B,num_coils,H,W,2) where num_coils is the number of coils.
+            """
+            return fftn_centered_t(complex_mul(x, sens_maps), spatial_dims=2)  # type: ignore
```
Comment (Can-Zhao) [inline monai/apps/reconstruction/networks/blocks/varnetblock.py:56]: Could you define the func outside `forward`? Thanks!

Comment (Can-Zhao) [reply monai/apps/reconstruction/networks/blocks/varnetblock.py:56]: Also it seems you hard-coded `spatial_dims=2`. Is any possibility that some people may use it for 3D?

File: monai/apps/reconstruction/networks/nets/coil_sensitivity_model.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            spatial_dims=spatial_dims,
+            features=features,
+            act=act,
+            norm=norm,
+            bias=bias,
+            dropout=dropout,
+            upsample=upsample,
+        )
+
+    def chans_to_batch_dim(self, x: Tensor) -> Sequence:
```
Comment (Can-Zhao) [inline monai/apps/reconstruction/networks/nets/coil_sensitivity_model.py:65]: Would it be better to spell out `chans` to `channels`? Thanks!

File: monai/apps/reconstruction/networks/nets/complex_unet.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+            out_channels=2,
+            features=features,
+            act=act,
+            norm=norm,
+            bias=bias,
+            dropout=dropout,
+            upsample=upsample,
+        )
+
+    def complex_to_chan_dim(self, x: Tensor) -> Tensor:
```
Comment (Can-Zhao) [inline monai/apps/reconstruction/networks/nets/complex_unet.py:67]: It seems it is hard-coded for 2D images

File: monai/apps/reconstruction/networks/nets/coil_sensitivity_model.py (hunk 3 above)
Comment (Can-Zhao) [reply monai/apps/reconstruction/networks/nets/coil_sensitivity_model.py:65]: Have you checked monai Reshape? https://github.com/Project-MONAI/MONAI/blob/31c4209f1d84a95536b56c3e4d6f7b5a31fe9b6f/monai/networks/layers/simplelayers.py#L140

File: monai/apps/reconstruction/networks/nets/complex_unet.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        Returns:
+            output of shape (B,C,H,W,2)
+        """
+        b, c2, h, w = x.shape  # c2 means c*2
+        if c2 % 2 != 0:
+            raise ValueError(f"channel dimension should be even but ({x.shape[1]}) is odd.")
+        c = c2 // 2
+        return x.view(b, 2, c, h, w).permute(0, 2, 3, 4, 1)
+
+    def normalize(self, x: Tensor) -> Sequence:
```
Comment (Can-Zhao) [inline monai/apps/reconstruction/networks/nets/complex_unet.py:99]: What is the difference between nn.GroupNorm?

Comment (Can-Zhao) [inline monai/apps/reconstruction/networks/nets/complex_unet.py:99]: What is the difference between LayerNorm?

File: monai/apps/reconstruction/networks/nets/complex_unet.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    def pad(self, x: Tensor) -> Sequence:
+        """
+        Pad input to feed into the network
+        """
+
+        def floor_ceil(n):
+            return math.floor(n), math.ceil(n)
+
+        b, c, h, w = x.shape
+        w_mult = ((w - 1) | 15) + 1
```
Comment (Can-Zhao) [inline monai/apps/reconstruction/networks/nets/complex_unet.py:144]: Could I ask how did 15 come from?

File: monai/apps/reconstruction/networks/nets/coil_sensitivity_model.py (hunk 3 above)
Comment (Can-Zhao) [reply monai/apps/reconstruction/networks/nets/coil_sensitivity_model.py:65]: Do you feel there will be other recon networks using this func? I was not sure. If you feel so, maybe put these funcs in `./nets/utils.py`.

File: monai/apps/reconstruction/networks/nets/complex_unet.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+
+        Returns:
+            output of shape (B,C*2,H,W)
+        """
+        b, c, h, w, two = x.shape
+        if x.shape[-1] != 2:
+            raise ValueError(f"last dim must be 2, but x.shape[-1] is {x.shape[-1]}.")
+        return x.permute(0, 4, 1, 2, 3).contiguous().view(b, 2 * c, h, w)
+
+    def chan_complex_to_last_dim(self, x: Tensor) -> Tensor:
```
Comment (Can-Zhao) [inline monai/apps/reconstruction/networks/nets/complex_unet.py:83]: Do you feel there will be other recon networks using this func? I was not sure. If you feel so, maybe add these funcs to `complex_utils.py`

File: monai/apps/reconstruction/networks/blocks/varnetblock.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+            Args:
+                x: kspace (B,num_coils,H,W,2) with the last dimension being 2 (for real/imaginary parts)
+
+            Returns:
+                reduction of x to (B,1,H,W,2)
+            """
+            x = ifftn_centered_t(x, spatial_dims=2)
+            return complex_mul(x, complex_conj(sens_maps)).sum(dim=1, keepdim=True)  # type: ignore
+
+        def soft_dc(x: Tensor) -> Tensor:
```
Comment (Can-Zhao) [inline monai/apps/reconstruction/networks/blocks/varnetblock.py:71]: Do you feel you may use this layer in the new network you design? If so, maybe you need to reconsider where to put it

File: monai/apps/reconstruction/networks/nets/complex_unet.py (hunk 5 above)
Comment (mersad95zd) [reply monai/apps/reconstruction/networks/nets/complex_unet.py]: Thanks for the link! According to the link, group norm is actually better because here features are divided into two groups (because of complexity of the data) and each group is normalized separately.

Comment (Can-Zhao) [reply monai/apps/reconstruction/networks/nets/complex_unet.py]: Yes, but your code seems to be Layer Norm...

File: monai/apps/reconstruction/networks/nets/utils.py [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+        d_mult = ((d - 1) | 15) + 1
+        w_pad = floor_ceil((w_mult - w) / 2)
+        h_pad = floor_ceil((h_mult - h) / 2)
+        d_pad = floor_ceil((d_mult - d) / 2)
+        x = F.pad(x, w_pad + h_pad + d_pad)
+        return x, (h_pad, w_pad, d_pad, h_mult, w_mult, d_mult)
+
+
+def reverse_pad(self, x: Tensor, pad_sizes: Sequence) -> Tensor:
+    """
```
Comment (Can-Zhao) [inline monai/apps/reconstruction/networks/nets/utils.py:166]: Could you help check whether `SpatialCrop()` may work here? https://github.com/Project-MONAI/MONAI/blob/46d9026c1ad5182ea8edc780771eb31929947b7c/monai/transforms/croppad/array.py#L465

Thanks

File: monai/apps/reconstruction/networks/nets/utils.py [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+    Returns:
+        A tuple containing:
+            (1) floor(n)
+            (2) ceil(n)
+    """
+    return math.floor(n), math.ceil(n)
+
+
+def pad(self, x: Tensor) -> Sequence:
+    """
```
Comment (Can-Zhao) [inline monai/apps/reconstruction/networks/nets/utils.py:133]: I was wondering iif the goal is to perform `DivisiblePad`. Could help check it? https://github.com/Project-MONAI/MONAI/blob/46d9026c1ad5182ea8edc780771eb31929947b7c/monai/transforms/croppad/array.py#L322

Thanks!

File: monai/apps/reconstruction/networks/nets/utils.py (hunk 9 above)
Comment (Can-Zhao) [reply monai/apps/reconstruction/networks/nets/utils.py:166]: Or may be the `inverse()` of `DivisiblePad` or `SpatialPad` might work?

File: monai/apps/reconstruction/networks/nets/complex_unet.py (hunk 5 above)
Comment (Can-Zhao) [reply monai/apps/reconstruction/networks/nets/complex_unet.py]: I would not mind whether you use Layer Norm or Group Norm. But please make sure the docstring matches with the code.

File: monai/apps/reconstruction/networks/nets/complex_unet.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+        spatial_dims: int = 2,
+        features: Sequence[int] = (32, 32, 64, 128, 256, 32),
+        act: Union[str, tuple] = ("LeakyReLU", {"negative_slope": 0.1, "inplace": True}),
+        norm: Union[str, tuple] = ("instance", {"affine": True}),
+        bias: bool = True,
+        dropout: Union[float, tuple] = 0.0,
+        upsample: str = "deconv",
+    ):
+        super().__init__()
+        self.unet = BasicUNet(
```
Comment (ericspod) [inline monai/apps/reconstruction/networks/nets/complex_unet.py:54]: I would suggest `ComplexUNet` should inherit from `BasicUNet` (or more ideally from `UNet`) instead of delegating like this so that inheritance can be used to change the behaviour of the class. Right now the internal structure of the class is hidden and immutable, it would be more flexible if inheritance could be used to override members.

File: monai/apps/reconstruction/networks/nets/complex_unet.py (hunk 5 above)
Comment (mersad95zd) [reply monai/apps/reconstruction/networks/nets/complex_unet.py]: oh I think I got your point.

File: monai/apps/reconstruction/networks/nets/utils.py [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+import math
+from typing import Sequence
+
+from torch import Tensor
+from torch.nn import functional as F
+
+
+
+def complex_to_channel_dim(self, x: Tensor) -> Tensor:
+    """
```
Comment (Can-Zhao) [inline monai/apps/reconstruction/networks/nets/utils.py:24]: I was wondering if it is better to name it as `reshape_complex_to_channel_dim` to indicate that this function reshape tensors. What do you think?

File: monai/apps/reconstruction/networks/nets/utils.py (hunk 10 above)
Comment (Can-Zhao) [reply monai/apps/reconstruction/networks/nets/utils.py:133]: Now I got it. It is not `DivisiblePad`. Would it be `SpatialPad`?

File: monai/apps/reconstruction/networks/nets/utils.py (hunk 12 above)
Comment (mersad95zd) [reply monai/apps/reconstruction/networks/nets/utils.py:24]: yeah, sure.

File: monai/apps/reconstruction/networks/nets/utils.py (hunk 10 above)
Comment (mersad95zd) [reply monai/apps/reconstruction/networks/nets/utils.py:133]: Yeah could be, I'll check whether SpatialPad can do this.

File: monai/apps/reconstruction/networks/nets/complex_unet.py (hunk 11 above)
Comment (mersad95zd) [reply monai/apps/reconstruction/networks/nets/complex_unet.py:60]: Thanks for pointing this out! Since I'll use the unet inside the forward function here, wouldn't it be easier and more readable to define it as a class attribute here?

File: monai/apps/reconstruction/networks/nets/utils.py [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+        mean: mean before normalization
+        std: std before normalization
+
+    Returns:
+        denormalized output of shape (B,C,H,W) for 2D data or (B,C,H,W,D) for 3D data
+    """
+    return x * std + mean
+
+
+def pad(x: Tensor) -> Sequence:  # type: ignore
```
Comment (Can-Zhao) [inline monai/apps/reconstruction/networks/nets/utils.py:122]: This function needs to be renamed... `pad()` is too general.

File: monai/apps/reconstruction/networks/nets/utils.py [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+
+    Args:
+        x: input of shape (B,C,H,W) for 2D data or (B,C,H,W,D) for 3D data
+        mean: mean before normalization
+        std: std before normalization
+
+    Returns:
+        denormalized output of shape (B,C,H,W) for 2D data or (B,C,H,W,D) for 3D data
+    """
+    return x * std + mean
```
Comment (Can-Zhao) [inline monai/apps/reconstruction/networks/nets/utils.py:119]: This func is only one line... Is it really needed?

File: monai/apps/reconstruction/networks/nets/complex_unet.py [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
```
Comment (Can-Zhao) [inline monai/apps/reconstruction/networks/nets/complex_unet.py:60]: I was wondering if it would be better to let self.unet here be replaced by an arbitrary model...So you could define `ComplexNet(net)`, which can take `net=BasicUnet()` or Resblock Unet or any network that meets some requirements. What do you think?

Comment (Can-Zhao) [reply monai/apps/reconstruction/networks/nets/complex_unet.py:60]: I realized self.unet has `in_channels=2, out_channels=2`. If you are going to let unet network be an input, this needs to be checked. It's up to you

File: monai/apps/reconstruction/networks/nets/utils.py (hunk 13 above)
Comment (Can-Zhao) [reply monai/apps/reconstruction/networks/nets/utils.py:122]: From the example, it looks to me that this is https://github.com/Project-MONAI/MONAI/blob/2c89142a14831a7f39d8c1acfaf1b83074aa041f/monai/transforms/croppad/array.py#L322

Could you double check it? Thanks!

Comment (mersad95zd) [reply monai/apps/reconstruction/networks/nets/utils.py]: Cool, I thought `DivisiblePad` was doing something different than our pad, but it's the same as you mentioned and I just fixed it.

**Review by Can-Zhao — APPROVED** [review]
Looks good to me.

File: monai/apps/reconstruction/networks/nets/complex_unet.py [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+
+    def forward(self, x: Tensor) -> Tensor:
+        """
+        Args:
+            x: input of shape (B,C,H,W,2) for 2D data or (B,C,H,W,D,2) for 3D data
+
+        Returns:
+            output of shape (B,C,H,W,2) for 2D data or (B,C,H,W,D,2) for 3D data
+        """
+        x = reshape_complex_to_channel_dim(x)
```
Comment (Can-Zhao) [inline monai/apps/reconstruction/networks/nets/complex_unet.py:84]: It would be helpful to add in-line comments to explain the shape of `x` for each operation on x.

File: monai/apps/reconstruction/networks/nets/complex_unet.py [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+        features: Sequence[int] = (32, 32, 64, 128, 256, 32),
+        act: Union[str, tuple] = ("LeakyReLU", {"negative_slope": 0.1, "inplace": True}),
+        norm: Union[str, tuple] = ("instance", {"affine": True}),
+        bias: bool = True,
+        dropout: Union[float, tuple] = 0.0,
+        upsample: str = "deconv",
+        pad_factor: int = 16,
+    ):
+        super().__init__()
+        self.unet = BasicUNet(
```
Comment (wyli) [inline monai/apps/reconstruction/networks/nets/complex_unet.py:63]: looks like this could be any fully convolutional network as long as the input and output channels are 2. if that's true, would be nice to make it a configurable part:

```py
class ComplexUnet(nn.Module):
    def __init__(
        ...,
        conv_net=None,
        ...
    ):
        if conv_net is None:
            self.unet = BasicUNet(...)
        else:
            self.unet = conv_net
        ...
```
